packages/plugin-hardhat/src/utils/impl-address.ts

```diff
@@ -0,0 +1,79 @@
+import { DeploymentNotFound, getBeaconAddress, getImplementationAddress, Manifest } from '@openzeppelin/upgrades-core';
+import { Contract, ethers, Signer, utils } from 'ethers';
+import { EthereumProvider, HardhatRuntimeEnvironment } from 'hardhat/types';
+import { ContractAddressOrInstance, getIBeaconFactory } from '.';
+
+/**
+ * Gets the implementation address from a Beacon.
+ *
+ * @returns the implementation address.
+ */
+export async function getImplementationAddressFromBeacon(
+  hre: HardhatRuntimeEnvironment,
+  signer: ethers.Signer | undefined,
+  beaconAddress: string,
+) {
+  const IBeaconFactory = await getIBeaconFactory(hre, signer);
+  const beaconContract = IBeaconFactory.attach(beaconAddress);
+  const currentImplAddress = await beaconContract.implementation();
+  return currentImplAddress;
+}
+
+/**
+ * Gets the implementation address from a UUPS/Transparent/Beacon proxy.
+ *
+ * @returns a Promise with the implementation address, or undefined if a UUPS/Transparent/Beacon proxy is not located at the address.
+ */
+export async function getImplementationAddressFromProxy(
+  provider: EthereumProvider,
+  proxyAddress: string,
+  hre: HardhatRuntimeEnvironment,
+  proxy: Contract | ContractAddressOrInstance,
+  signer: Signer | undefined,
+) {
+  let result: string | undefined;
+  try {
+    result = await getImplementationAddress(provider, proxyAddress);
+  } catch (e: any) {
+    try {
+      const beaconAddress = await getBeaconAddress(provider, proxyAddress);
+      result = await getImplementationAddressFromBeacon(
+        hre,
+        proxy instanceof Contract ? proxy.signer : signer,
+        beaconAddress,
+      );
+    } catch (e: any) {
+      // error expected if the address was not a beacon proxy
```

Comment:
Same as a previous comment, we shouldn't ignore the error if it's a network error or something else.